### PR TITLE
(PUP-10473) Remove legacy auth.conf

### PR DESCRIPTION
**Pull request by joshcooper**
[Puppetserver's auth.conf](https://puppet.com/docs/puppetserver/latest/config_file_auth.html) controls access to all of puppet's REST APIs `/puppet/v3` and `/puppet-ca/v1`. However, if the `use-legacy-auth-conf` setting is true, then puppetserver will delegate to puppet to perform authz checks for `/puppet/v3` only using its [legacy auth.conf](https://puppet.com/docs/puppet/latest/config_file_auth.html). This includes requests to static APIs like `/puppet/v3/catalog` and custom file mounts defined in [fileserver.conf](https://puppet.com/docs/puppet/latest/config_file_fileserver.html).

Puppet's legacy auth.conf has been deprecated since 4.x. Additionally, it does not support authorizing requests based on certificate extensions, which is something that puppetserver's auth.conf does, and is needed to control access to privileged API endpoints.

This PR removes:
* Authorization checks for indirected, environment and custom fileserver mounts
* Legacy auth implementation
* `rest_authconfig` setting
* Default `auth.conf` which [puppet-agent used to include at package time](https://github.com/puppetlabs/puppet-agent/commit/ae177a0ef5f62b75bbf654b271eb78ab4a0db32c).

This PR preserves the `Puppet::Network::AuthConfig.authprovider_class=` and `Puppet::Network::Authorization.authconfigloader_class=` setters, but makes them noop. This is needed because puppetserver calls these hooks to register its noop authorization provider, so that puppet always allows requests, giving puppetserver a chance to handle them.

Once this PR is merged and a new puppet-agent build promoted, then puppetserver can be modified to stop calling the above methods. And then they can be be removed from puppet.

### CODE REVIEW

**Comment by joshcooper** [comment]
jenkins please test this with servertests

**Comment by joshcooper** [comment]
jenkins please test this with servertests

**Comment by joshcooper** [comment]
Well that's annoying, the CI check is still showing the first failure, while the actual run is progressing in https://jenkins-platform.delivery.puppetlabs.net/job/platform_puppet-agent-extra_puppet-agent-integration-suite_pr/1844/

**Comment by joshcooper** [comment]
I need to rebase this after the app orchestration PR lands due to the `/puppet/v3/environment` route changes.

**Comment by joshcooper** [comment]
jenkins please test this with servertests

**Comment by joshcooper** [comment]
This is blocked on https://github.com/puppetlabs/puppet-agent/pull/1970

**Comment by joshcooper** [comment]
jenkins please test this on redhat7-64a,ubuntu1804a,windows2019-64a with servertests

**Comment by joshcooper** [comment]
jenkins please test this on redhat7-64a,ubuntu1804-64a,windows2019-64a with servertests

**Comment by joshcooper** [comment]
This is blocked on https://github.com/puppetlabs/puppetserver/pull/2416

**Comment by joshcooper** [comment]
jenkins please test this on redhat7-64a,ubuntu1804-64a,windows2019-64a with servertests

File: lib/puppet/file_serving/configuration.rb
Comment (mwaggett) [inline lib/puppet/file_serving/configuration.rb:83]: What were these for?

File: spec/unit/file_serving/configuration/parser_spec.rb
Comment (mwaggett) [inline spec/unit/file_serving/configuration/parser_spec.rb:111]: Are these testing anything anymore at all? Or can they be removed completely?

File: spec/unit/file_serving/configuration_spec.rb
Comment (mwaggett) [inline spec/unit/file_serving/configuration_spec.rb:111]: What were these for?

**Review by mwaggett — APPROVED** [review]
This looks sane to me, just a couple questions since this is new to me.
Would probably be good to get someone more familiar to also take a look.

File: conf/fileserver.conf
Comment (joshcooper) [reply conf/fileserver.conf:15]: The default `fileserver.conf` that we installed made it seem like you could specify allow/deny rules, but in reality you couldn't. Trying to restrict access to something less than "allow *" would result in access denied due to https://github.com/puppetlabs/puppet/pull/8320/commits/5ffd9f37a20ab6a7f760bb46f469a20aa11e0215#diff-68f0ad319ef3f6122b7e02a74ba137daL21-L26 with a message about using auth.conf instead. Now that everything is controlled via puppetserver's auth.conf, I thought it best to remove this confusion.

File: lib/puppet/file_serving/configuration.rb
Comment (joshcooper) [reply lib/puppet/file_serving/configuration.rb:83]: Related to above, previously the hard coded mounts had to be global allowed, on the off chance that the legacy auth was enabled. Now that puppetserver's auth.conf performs the check, we no longer need to store allow/deny rules for mounts.

File: spec/unit/file_serving/configuration/parser_spec.rb
Comment (joshcooper) [reply spec/unit/file_serving/configuration/parser_spec.rb:111]: I left them in to make sure we still parse `fileserver.conf` with allow/deny rules in them, as could occur during an upgrade. Currently we log an error and continue. But I should probably update the test to expect the error message.

I also considering having the parser raise if it encountered an allow or deny rule. But I feel like that would bit us during upgrades.

File: spec/unit/file_serving/configuration_spec.rb
Comment (joshcooper) [reply spec/unit/file_serving/configuration_spec.rb:111]: This was to stub out the call to [Mount#empty?](https://github.com/puppetlabs/puppet/pull/8320/commits/5ffd9f37a20ab6a7f760bb46f469a20aa11e0215#diff-68f0ad319ef3f6122b7e02a74ba137daL21). But now we don't store rules for each mount, just the name of the mount and the local `path` where the mount maps to.

File: spec/unit/file_serving/configuration/parser_spec.rb
Comment (mwaggett) [reply spec/unit/file_serving/configuration/parser_spec.rb:111]: ahh, yeah, I think it would be clearer to future us if you added an expectation.

File: api/docs/http_api_index.md
Comment (joshcooper) [reply api/docs/http_api_index.md:36]: Maybe something like?
> As of Puppet 7, support for legacy `auth.conf` is removed. Puppet Server 7 authorizes access to its REST endpoints solely using its `auth.conf` and `.ca.conf`. See https://puppet.com/docs/puppetserver/latest/config_file_auth.html for more details.

File: api/docs/http_certificate_request.md
Comment (joshcooper) [reply api/docs/http_certificate_request.md:54]: I'm not seeing those docs on the docs website, e.g. https://puppet.com/docs/puppetserver/latest/index.html doesn't show the CA APIs. I can file a ticket to move/remove puppet docs once the puppetserver docs migration is done.

Comment (Magisus) [reply api/docs/http_certificate_request.md:54]: We have these in the side nav https://puppet.com/puppetserver/6.3/http_certificate_request.html but they 404 :(

File: spec/unit/file_serving/configuration/parser_spec.rb
Comment (joshcooper) [reply spec/unit/file_serving/configuration/parser_spec.rb:111]: Ah so I kept the tests above to show that we could still parse the same inputs as before and that the `acl_type` (either `allow` or `deny` is ignored). The tests after those that [verify the error is logged as expected](https://github.com/puppetlabs/puppet/pull/8320/files#diff-483269bc1e381cfdc995bc8e7b421b89R119).

File: api/docs/http_api_index.md
Comment (Magisus) [reply api/docs/http_api_index.md:36]: I think you can drop the `and ca.conf`. That's not used for authorization.

Comment (joshcooper) [reply api/docs/http_api_index.md:36]: ahhhh, right, I was conflating those two.

File: acceptance/tests/allow_arbitrary_node_name_fact_for_agent.rb
Comment (joshcooper) [reply acceptance/tests/allow_arbitrary_node_name_fact_for_agent.rb]: Updated

File: acceptance/tests/allow_arbitrary_node_name_for_agent.rb
Comment (joshcooper) [reply acceptance/tests/allow_arbitrary_node_name_for_agent.rb]: Updated

File: spec/unit/file_serving/configuration/parser_spec.rb
Comment (mwaggett) [reply spec/unit/file_serving/configuration/parser_spec.rb:111]: I still think it looks weird to have tests that don't have any expectations.. 
Maybe just `expect { @parser.parse }.not_to raise_error`?

Comment (joshcooper) [reply spec/unit/file_serving/configuration/parser_spec.rb:111]: Expecting the test not to raise is a bit redundant (since rspec will fail the test if it does). But to your point, it'd be clearer to expect that `@parser.parsrer` returns the expected mounts.

File: api/docs/http_certificate_request.md
Comment (joshcooper) [reply api/docs/http_certificate_request.md:54]: Also puppet's API docs still reference legacy routes, so probably best to make a pass through the docs after the dust settles with the website and legacy routes changes.

File: spec/unit/file_serving/configuration/parser_spec.rb
Comment (joshcooper) [reply spec/unit/file_serving/configuration/parser_spec.rb:111]: Updated

**Review by Magisus — APPROVED** [review]

File: spec/unit/file_serving/configuration/parser_spec.rb
Comment (mwaggett) [reply spec/unit/file_serving/configuration/parser_spec.rb:111]: I didn't see any update.. Is my stuff out of date or did you forget to push?

Comment (joshcooper) [reply spec/unit/file_serving/configuration/parser_spec.rb:111]: Oh jeez... looks like I didn't push... The diff is just the test changes, so I'll submit a different PR for that:

```diff
$ git diff 2d60297d00..c64418b370
diff --git a/spec/unit/file_serving/configuration/parser_spec.rb b/spec/unit/file_serving/configuration/parser_spec.rb
index 72116de7a5..32dbfd3268 100644
--- a/spec/unit/file_serving/configuration/parser_spec.rb
+++ b/spec/unit/file_serving/configuration/parser_spec.rb
@@ -101,13 +101,14 @@ describe Puppet::FileServing::Configuration::Parser do
       it "should ignore inline comments in #{acl_type}" do
         write_config_file "[one]\n#{acl_type} something \# will it work?\n"
 
-        @parser.parse
+        expect(@parser.parse).to eq('one' => @mount)
+
       end
 
       it "should ignore #{acl_type} from ACLs with varying spacing around commas" do
         write_config_file "[one]\n#{acl_type} someone,sometwo, somethree , somefour ,somefive\n"
 
-        @parser.parse
+        expect(@parser.parse).to eq('one' => @mount)
       end
 
       it "should log an error and print the location of the #{acl_type} rule" do
```

**Pull request merged**
